File: ontrack-service/src/main/java/net/nemerosa/ontrack/service/security/ConfidentialStore.java
package net.nemerosa.ontrack.service.security;

import java.io.IOException;

public interface ConfidentialStore {
    /**
     * Persists the payload of {@link ConfidentialKey} to a persisted storage (such as disk.)
     * The expectation is that the persisted form is secure.
     */
    void store(ConfidentialKey key, byte[] payload) throws IOException;

    /**
     * Reverse operation of {@link #store(ConfidentialKey, byte[])}
     *
     * @return null the data has not been previously persisted, or if the data was tampered.
     */
    byte[] load(ConfidentialKey key) throws IOException;

    /**
     * Works like {@link java.security.SecureRandom#nextBytes(byte[])}.
     * <p>
     * This enables implementations to consult other entropy sources, if it's available.
     */
    byte[] randomBytes(int size);

    /**
     * Gets the unique ID for this store, which allows for its selection in
     * {@link net.nemerosa.ontrack.model.support.OntrackConfigProperties}.
     *
     * @return A unique ID
     * @see
     */
    String getId();
}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/security/ConfidentialStore.kt
package net.nemerosa.ontrack.model.security

interface ConfidentialStore {
    /**
     * Persists the payload of key to a persisted storage (such as disk.)
     * The expectation is that the persisted form is secure.
     */
    fun store(key: String, payload: ByteArray)

    /**
     * Reverse operation of [store]
     *
     * @return null the data has not been previously persisted, or if the data was tampered.
     */
    fun load(key: String): ByteArray?

    /**
     * Works like [java.security.SecureRandom.nextBytes].
     *
     * This enables implementations to consult other entropy sources, if it's available.
     */
    fun randomBytes(size: Int): ByteArray
}